oshmem: Use the right datatype for PMIX_DATA_BUFFER_UNLOAD

rzambre opened the pull request:
Fixes segfaults during shmem_init() reported in #11430 and #11524.

jsquyres commented on oshmem/mca/memheap/base/memheap_base_mkey.c:
```diff
@@ -445,14 +445,14 @@ void memheap_oob_destruct(void)
     free(buffer);
 
-    MEMHEAP_VERBOSE(5, "message sent: dst=%d, rc=%d, %d bytes!", pe, rc, size);
+    MEMHEAP_VERBOSE(5, "message sent: dst=%d, rc=%d, %zu bytes!", pe, rc, size);
```
Please use the appropriate `PRI` printf code for portability: https://github.com/open-mpi/ompi/wiki/PrintfCodes

jsquyres commented on oshmem/mca/memheap/base/memheap_base_mkey.c:
```diff
@@ -584,7 +584,7 @@ void mca_memheap_modex_recv_all(void)
     /* Do allgather */
     PMIX_DATA_BUFFER_UNLOAD(msg, send_buffer, size);
-    MEMHEAP_VERBOSE(1, "local keys packed into %d bytes, %d segments", size, memheap_map->n_segments);
+    MEMHEAP_VERBOSE(1, "local keys packed into %d bytes, %zu segments", size, memheap_map->n_segments);
```
Same here.

jsquyres commented:
@janjust Ping

jsquyres commented:
@rzambre While you've got the hood up, could you clean up some other issues in OSHMEM?  In just test compiling the `memheap` tree, I notice a bunch of other warnings.  Could you fix all of these warnings?  Thanks!

```
$ make
  CC       base/memheap_base_frame.lo
base/memheap_base_frame.c:38:1: warning: braces around scalar initializer
 mca_memheap_map_t mca_memheap_base_map = {{{{0}}}};
 ^~~~~~~~~~~~~~~~~
base/memheap_base_frame.c:38:1: note: (near initialization for ‘mca_memheap_base_map.mem_segs’)
base/memheap_base_frame.c:38:1: warning: braces around scalar initializer
base/memheap_base_frame.c:38:1: note: (near initialization for ‘mca_memheap_base_map.mem_segs’)
base/memheap_base_frame.c:38:1: warning: braces around scalar initializer
base/memheap_base_frame.c:38:1: note: (near initialization for ‘mca_memheap_base_map.mem_segs’)
  CC       base/memheap_base_select.lo
  CC       base/memheap_base_alloc.lo
  CC       base/memheap_base_static.lo
  CC       base/memheap_base_register.lo
base/memheap_base_register.c: In function ‘mca_memheap_base_reg’:
base/memheap_base_register.c:24:45: warning: declaration of ‘memheap_map’ shadows a global declaration [-Wshadow]
 int mca_memheap_base_reg(mca_memheap_map_t *memheap_map)
                          ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
In file included from base/memheap_base_register.c:17:
../../../oshmem/mca/memheap/base/base.h:164:27: note: shadowed declaration is here
 extern mca_memheap_map_t* memheap_map;
                           ^~~~~~~~~~~
base/memheap_base_register.c: In function ‘mca_memheap_base_dereg’:
base/memheap_base_register.c:50:47: warning: declaration of ‘memheap_map’ shadows a global declaration [-Wshadow]
 int mca_memheap_base_dereg(mca_memheap_map_t *memheap_map)
                            ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
In file included from base/memheap_base_register.c:17:
../../../oshmem/mca/memheap/base/base.h:164:27: note: shadowed declaration is here
 extern mca_memheap_map_t* memheap_map;
                           ^~~~~~~~~~~
  CC       base/memheap_base_mkey.lo
In file included from /home/jsquyres/git/o2/3rd-party/openpmix/include/pmix_common.h:2647,
                 from /home/jsquyres/git/o2/3rd-party/openpmix/include/pmix.h:53,
                 from ../../../opal/mca/pmix/pmix-internal.h:47,
                 from ../../../opal/util/proc.h:28,
                 from ../../../oshmem/proc/proc.h:23,
                 from base/memheap_base_mkey.c:22:
base/memheap_base_mkey.c: In function ‘send_buffer’:
/home/jsquyres/git/o2/3rd-party/openpmix/include/pmix_deprecated.h:851:32: warning: passing argument 2 of ‘PMIx_Data_buffer_unload’ from incompatible pointer type [-Wincompatible-pointer-types]
     PMIx_Data_buffer_unload(b, &(d), &(s))
                                ^~~~
base/memheap_base_mkey.c:451:5: note: in expansion of macro ‘PMIX_DATA_BUFFER_UNLOAD’
     PMIX_DATA_BUFFER_UNLOAD(msg, buffer, size);
     ^~~~~~~~~~~~~~~~~~~~~~~
/home/jsquyres/git/o2/3rd-party/openpmix/include/pmix_deprecated.h:352:49: note: expected ‘char **’ but argument is of type ‘void **’
                                          char **bytes, size_t *sz);
                                          ~~~~~~~^~~~~
base/memheap_base_mkey.c: In function ‘mca_memheap_modex_recv_all’:
/home/jsquyres/git/o2/3rd-party/openpmix/include/pmix_deprecated.h:851:32: warning: passing argument 2 of ‘PMIx_Data_buffer_unload’ from incompatible pointer type [-Wincompatible-pointer-types]
     PMIx_Data_buffer_unload(b, &(d), &(s))
                                ^~~~
base/memheap_base_mkey.c:586:5: note: in expansion of macro ‘PMIX_DATA_BUFFER_UNLOAD’
     PMIX_DATA_BUFFER_UNLOAD(msg, send_buffer, size);
     ^~~~~~~~~~~~~~~~~~~~~~~
/home/jsquyres/git/o2/3rd-party/openpmix/include/pmix_deprecated.h:352:49: note: expected ‘char **’ but argument is of type ‘void **’
                                          char **bytes, size_t *sz);
                                          ~~~~~~~^~~~~
  CCLD     libmca_memheap.la
```

rzambre commented:
@jsquyres I've taken note of the warnings and I will be happy to address them in a follow-up PR. I addressed your comments immediately since the changes in this PR are rather critical. Without them, OSHMEM is broken.
